packages/connect/src/relay/context.ts

```diff
@@ -327,149 +442,193 @@ class RelayContext extends EventEmitter {
+
+    // On every reconnect, create a new asyncIterable to pass into
+    async function* drain(
+      this: RelayContext,
+      internalIteration: number,
+      endPromise: DeferType<SinkEndedEvent | EndedEvent>
+    ) {
+      const advanceIterator = () => {
+        if (currentSource == null) {
+          throw Error(`Source is not yet set`)
+        }
+        nextMessagePromise = currentSource.next().then((res) => ({
+          type: ConnectionEventTypes.PAYLOAD,
+          value: res
+        }))
       }
 
-      let ended = false
-      const endPromise = end.promise.then(() => {
-        ended = true
-      })
-
+      // Notify *why* stream has ended
+      let reasonToLeave: ConnectionEventTypes.STREAM_ENDED | ConnectionEventTypes.ENDED =
+        ConnectionEventTypes.STREAM_ENDED
+      let leave = false
       this.flow(`FLOW: relay_outgoing: loop started`)
-      while (true) {
-        this.flow(`FLOW: relay_outgoing: new loop iteration`)
-
-        const promises: Promise<SinkResult>[] = []
 
-        let resolvedPromiseName
-
-        const pushPromise = (promise: Promise<any>, name: string) => {
-          promises.push(
-            promise.then((res) => {
-              resolvedPromiseName = name
-              return res
-            })
-          )
-        }
+      while (!leave) {
+        this.flow(`FLOW: relay_outgoing: new loop iteration`)
 
-        pushPromise(endPromise, 'ended')
+        const promises: Promise<SinkEvent>[] = []
 
         if (currentSource == undefined) {
-          pushPromise(this._sinkSourceAttachedPromise.promise, 'sinkSourceAttacked')
+          promises.push(this._sinkSourceAttachedPromise.promise)
         }
 
-        pushPromise(this._statusMessagePromise.promise, 'statusMessage')
+        promises.push(this._statusMessagePromise.promise)
 
-        if (currentSource != undefined && (result == undefined || (result as StreamResult).done != true)) {
-          sourcePromise = sourcePromise ?? currentSource.next()
+        if (currentSource != undefined) {
+          nextMessagePromise =
+            nextMessagePromise ??
+            currentSource.next().then((res) => ({
+              type: ConnectionEventTypes.PAYLOAD,
+              value: res
+            }))
 
-          pushPromise(sourcePromise, 'payload')
+          promises.push(nextMessagePromise)
         }
 
-        // (0. Handle source attach)
-        // 1. Handle stream switches
-        // 2. Handle status messages
-        // 3. Handle payload messages
         this.flow(`FLOW: relay_outgoing: awaiting promises`)
-        result = await Promise.race(promises)
-
-        this.flow(`FLOW: relay_outgoing: promise ${resolvedPromiseName} resolved`)
+        const result = await Promise.race(promises)
 
-        // Don't handle incoming messages after migration
-        if (ended) {
-          this.flow(`FLOW: relay_outgoing: ended, break`)
+        if (iteration != internalIteration) {
+          leave = true
```

Comment:
Might missed it before, but `leave = true` is useless here due to the `break` on the next line.